[HUDI-151] Enable HiveOnSpark queries for RT tables

n3nash opened the pull request:
- Adding more comments to understand usage of reader/writer schema
- Fixing RT queries for HiveOnSpark that causes race conditions

vinothchandar commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/HoodieParquetRealtimeInputFormat.java:
```diff
@@ -186,7 +187,7 @@ private static Configuration addProjectionField(Configuration conf, String field
   }
 
-  private static synchronized Configuration addRequiredProjectionFields(Configuration configuration) {
+  private Configuration addRequiredProjectionFields(Configuration configuration) {
```
can we keep this static, if it does not need any state from the instance?

vinothchandar commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/HoodieParquetRealtimeInputFormat.java:
```diff
@@ -203,14 +204,12 @@ private static synchronized Configuration addRequiredProjectionFields(Configurat
-        if (LOG.isDebugEnabled()) {
-          LOG.debug("The projection Ids: {" + columnIds + "} start with ','. First comma is removed");
-        }
+  private Configuration cleanProjectionColumnIds(Configuration conf) {
```
keep it static?

n3nash commented:
@vinothchandar Addressed your comment, also added change to another class - this was pushed as part of a different PR earlier and then reverted so adding it back since it is required.
@bvaradar please review

n3nash commented:
@vinothchandar @bvaradar Addressed your comments and made some more changes, please take another pass before we can land this.

vinothchandar commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/AbstractRealtimeRecordReader.java:
```diff
@@ -156,24 +158,46 @@ protected static String arrayWritableToString(ArrayWritable writable) {
+    String[] fieldOrdersWithDups = fieldOrderCsv.split(",");
+    List<String> fieldOrdersList = new LinkedList<>();
+    for (String fieldOrder : fieldOrdersWithDups) {
+      if (!fieldOrdersSet.contains(fieldOrder)) {
```
if you are using the set for de-duping and the list for order preservation,may be just use a https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashSet.html ? and simplify?

vinothchandar commented on hudi-hadoop-mr/src/main/java/org/apache/hudi/hadoop/realtime/AbstractRealtimeRecordReader.java:
```diff
@@ -156,24 +158,46 @@ protected static String arrayWritableToString(ArrayWritable writable) {
+    List<String> fieldNames = Arrays.stream(fieldNameCsv.split(","))
+        .filter(fn -> !partitioningFields.contains(fn)).collect(Collectors.toList());
+    Set<String> fieldNamesSet = new HashSet<>();
+    List<String> fieldNamesList = new LinkedList<>();
```
same here.

bvaradar commented:
@n3nash : Once you address the comment, I will merge this diff

n3nash commented:
@bvaradar Guessing you are referring to the build, should be fixed now, feel free to merge
